File: webapp/apps/regolia/src/models/entities/course/semester-course.ts
import {Entity} from "../entity";
import {Course} from "./course.entity";
import {SemesterLevel} from "../semester";
import {SemesterCourseTeacher} from "./semester-course-teacher";
import {SemesterCourseLevelSpeciality} from "./semester-course-level-speciality";
import {List} from "@positon/collections";
import {CourseHour} from "./course-hour.entity";
import {CourseSession} from "./course-session.entity";

export class SemesterCourse extends Entity<number> {
  constructor(value: any = {}) {
    super();

    this.id = value.id;
    this.registrationDate = new Date(value.registrationDate);

    this.practicalWork = value.practicalWork;
    this.isGeneral = value.isGeneral;
    this.radical = value.radical;
    this.coefficient = value.coefficient;
    this.courseId = value.courseId;
    this.semesterLevelId = value.semesterLevelId;

    if(value.course) {
      this.course = new Course(value.course);
    }

    if(value.semesterCourseLevelSpecialities) {
      this.semesterCourseLevelSpecialities = value.semesterCourseLevelSpecialities
        .map(u => new SemesterCourseLevelSpeciality(u) );

      this.semesterCourseLevelSpecialities.forEach(s => s.semesterCourse = this);
    }
  }

  coefficient: boolean;
  practicalWork: boolean;
  isGeneral: boolean;
  radical: number;

  course: Course;
  courseId: number;

  semesterLevel: SemesterLevel;
  semesterLevelId: number;

  semesterCourseLevelSpecialities: SemesterCourseLevelSpeciality[];

  semesterCourseTeachers: SemesterCourseTeacher[];

  courseHours: List<CourseHour>;
  courseSessions: List<CourseSession>;

  get specialityNames(): string[] {
    return this.semesterCourseLevelSpecialities
      .map(s => s.semesterLevelSpeciality?.yearLevelSpeciality?.levelSpeciality?.speciality.name)
  }

  url(path?: string): string {
    const url = `${this.semesterLevel?.url()}/courses/${this.id}`;
    if(path) {
      return `${url}/${path}`;
    }
    return url;
  }
}

File: ServerApp/Entities/Periods/Semester/SemesterCourse.cs
using System.Collections.Generic;
using Everest.AspNetStartup.Binding;
using Everest.AspNetStartup.Models;
using Exam.Entities.Courses;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;

namespace Exam.Entities.Periods
{
    [ModelBinder(BinderType = typeof(ItemValueModelBinder))]
    public class SemesterCourse:Entity<long>
    {
        
        public bool PracticalWork { get; set; }
        
        public bool IsGeneral { get; set; }
        
        public uint Coefficient { get; set; }
        
        public uint Radical { get; set; }

        public bool Required { get; set; } = true;

        public virtual Course Course { get; set; }
        public long? CourseId { get; set; }

        [JsonIgnore]
        public virtual SemesterLevel SemesterLevel { get; set; }
        public long? SemesterLevelId { get; set; }

        
        public virtual List<SemesterCourseLevelSpeciality> SemesterCourseLevelSpecialities { get; set; }
        
        [JsonIgnore]
        public virtual List<CourseHour> CourseHours { get; set; }
        
        [JsonIgnore]
        public virtual List<CourseSession> CourseSessions { get; set; }
        
        [JsonIgnore] public virtual List<Test> Tests { get; set; }
        
        
        
    }
}